crates/rpc-types-engine/src/superchain.rs

```diff
@@ -0,0 +1,180 @@
+    InvalidLength(usize),
+    #[error("Invalid version format encoding")]
+    FromSlice(#[from] TryFromSliceError),
+}
+
+impl From<ProtocolVersion> for B256 {
+    fn from(value: ProtocolVersion) -> B256 {
+        let mut bytes = [0u8; 32];
+
+        match value {
+            ProtocolVersion::V0(value) => {
+                // <protocol version> ::= <version-type><typed-payload>
+                // <version-type> ::= <uint8>
+                // <typed-payload> ::= <31 bytes>
+                bytes[0] = 0x00; // this is not necessary, but addded for clarity
+                bytes[1..].copy_from_slice(&value.into_slice());
+                B256::from_slice(&bytes)
+            }
+        }
+    }
+}
+
+impl TryFrom<B256> for ProtocolVersion {
+    type Error = ProtocolVersionError;
+
+    fn try_from(value: B256) -> Result<Self, Self::Error> {
+        // <protocol version> ::= <version-type><typed-payload>
+        // <version-type> ::= <uint8>
+        // <typed-payload> ::= <31 bytes>
+        let version_type = value[0];
+        let typed_payload = &value[1..];
+
+        match version_type {
+            0 => Ok(Self::V0(ProtocolVersionFormatV0::try_from_slice(typed_payload)?)),
+            other => Err(ProtocolVersionError::UnsupportedVersion(other)),
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl serde::Serialize for ProtocolVersion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        B256::from(*self).serialize(serializer)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for ProtocolVersion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = B256::deserialize(deserializer)?;
+        Self::try_from(value).map_err(serde::de::Error::custom)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ProtocolVersionFormatV0 {
+    pub build: B64,
+    pub major: u32,
+    pub minor: u32,
+    pub patch: u32,
+    pub pre_release: u32,
+}
+
+impl std::fmt::Display for ProtocolVersionFormatV0 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{build}.{major}.{minor}.{patch}-{pre_release}",
+            build = self.build,
+            major = self.major,
+            minor = self.minor,
+            patch = self.patch,
+            pre_release = self.pre_release
+        )
+    }
+}
+
+impl ProtocolVersionFormatV0 {
+    /// Version-type 0 byte encoding:
+    ///
+    /// ```text
+    /// <reserved><build><major><minor><patch><pre-release>
+    /// <reserved> ::= <7 zeroed bytes>
+    /// <build> ::= <8 bytes>
+    /// <major> ::= <big-endian uint32>
+    /// <minor> ::= <big-endian uint32>
+    /// <patch> ::= <big-endian uint32>
+    /// <pre-release> ::= <big-endian uint32>
+    /// ```
+    pub fn into_slice(self) -> [u8; 31] {
```

Comment:
this is not a slice.

perhaps just encode?

Reply:
renamed to `encode` and `decode` respectively